esp-hal/src/spi/master.rs

```diff
@@ -1140,48 +1420,79 @@ mod dma {
                 // of the received data as possible into the buffer and
                 // discarded the rest. The user doesn't care about this discarded data.
 
-                if !self.spi_dma.channel.rx.is_done()
-                    && !self.spi_dma.channel.rx.has_dscr_empty_error()
+                if !self.spi_dma.channel().rx.is_done()
+                    && !self.spi_dma.channel().rx.has_dscr_empty_error()
                 {
                     return false;
                 }
             }
             true
         }
 
+        fn do_wait(&self) {
+            while !self.is_done() {
+                // Wait for the transfer to complete
+            }
+            fence(Ordering::Acquire);
+        }
+
         /// Waits for the DMA transfer to complete.
         ///
         /// This method blocks until the transfer is finished and returns the
         /// `SpiDma` instance and the associated buffer.
-        pub fn wait(self) -> (SpiDma<'d, T, C, D, M>, Buf) {
-            while !self.is_done() {
-                // Wait for the transfer to complete
+        pub fn wait(mut self) -> (R, Buf) {
+            self.do_wait();
+            unsafe {
+                (
+                    ManuallyDrop::take(&mut self.spi_dma),
+                    ManuallyDrop::take(&mut self.dma_buf),
+                )
+            }
+        }
+
+        /// Cancels the DMA transfer.
+        pub fn cancel(&mut self) {
```

Comment:
You cancelled the DMA but you haven't cancelled the SPI? The transfer will still continue in the background, it's just that garbage would be sent out instead.

Reply:
Yeah apparently only on the S2 does the SPI actually stop...

Reply:
I'm utterly shocked that it actually stopped, even if it's just the S2.

The "workaround" I had for this was going to be some kind of `PhantomTransfer` type, where the DMA stops running (allowing you to get back ownership of the dma buf) but the SPI continues in the background.